File: code-generator/src/main/java/io/github/potjerodekool/openapi/internal/di/AnnotationInvocationHandler.java
package io.github.potjerodekool.openapi.internal.di;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.util.Map;
import java.util.StringJoiner;

class AnnotationInvocationHandler implements InvocationHandler {

    private final Class<?> annotationClass;

    private final Map<String, Object> attributes;

    public AnnotationInvocationHandler(final Class<?> annotationClass,
                                       final Map<String, Object> attributes) {
        this.annotationClass = annotationClass;
        this.attributes = attributes;
    }

    @Override
    @SuppressWarnings("return")
    public Object invoke(final Object proxy, final Method method, final Object[] args) throws Throwable {
        final var methodName = method.getName();

        if (method.isDefault()) {
            return InvocationHandler.invokeDefault(proxy, method, args);
        }

        return switch (methodName) {
            case "annotationType" -> annotationClass;
            case "equals" -> this.equals(args[0]);
            case "hashCode" -> this.hashCode();
            case "toString" -> this.toString();
            default -> {
                if (method.getReturnType() == Void.class || method.getReturnType() == Void.TYPE) {
                    yield null;
                } else if (method.getParameterCount() == 0) {
                    if (this.attributes.containsKey(methodName)) {
                        yield this.attributes.get(methodName);
                    }
                }

                yield null;
            }
        };
    }

    public String toString() {
        final var sb = new StringBuilder();
        sb.append("@");
        sb.append(annotationClass.getName());
        sb.append("(");

        final var attributeJoiner = new StringJoiner(",");

        attributes.forEach((key, value) -> attributeJoiner.add(quoteString(key) + "=" + quoteString(value)));

        sb.append(attributeJoiner);
        sb.append(")");

        return sb.toString();
    }

    private Object quoteString(final Object value) {
        if (value instanceof String) {
            return "\"" + value + "\"";
        } else {
            return value;
        }
    }
}

File: code-generator/src/main/kotlin/io/github/potjerodekool/openapi/internal/di/AnnotationInvocationHandler.kt
package io.github.potjerodekool.openapi.internal.di

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.util.*

internal class AnnotationInvocationHandler(
    private val annotationClass: Class<*>,
    private val attributes: Map<String, Any>
) : InvocationHandler {
    @Throws(Throwable::class)
    override fun invoke(proxy: Any, method: Method, args: Array<Any>?): Any? {
        val methodName = method.name

        if (method.isDefault) {
            return InvocationHandler.invokeDefault(proxy, method, args)
        }

        return when (methodName) {
            "annotationType" -> annotationClass
            "equals" -> !args.isNullOrEmpty() && this == args[0]
            "hashCode" -> this.hashCode()
            "toString" -> this.toString()
            else -> {
                if (method.returnType == Void::class.java || method.returnType == Void.TYPE) {
                    return null
                } else if (method.parameterCount == 0) {
                    if (attributes.containsKey(methodName)) {
                        return attributes[methodName]
                    }
                }
                null
            }
        }!!
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("@")
        sb.append(annotationClass.name)
        sb.append("(")

        val attributeJoiner = StringJoiner(",")

        attributes.forEach { (key: String, value: Any) ->
            attributeJoiner.add(
                quoteString(key).toString() + "=" + quoteString(
                    value
                )
            )
        }

        sb.append(attributeJoiner)
        sb.append(")")

        return sb.toString()
    }

    private fun quoteString(value: Any): Any {
        return if (value is String) {
            "\"" + value + "\""
        } else {
            value
        }
    }
}
